Module tag app aware search

Jtang-1 opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
non user facing changes. Adds `module_name` tag to Datadog metric `commcare.app_aware_search.processing_time` 

![Screen Shot 2024-03-08 at 4 31 03 PM](https://github.com/dimagi/commcare-hq/assets/88759246/96cb64b0-80a4-4139-944f-ce169299ce17)

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Jira Ticket: https://dimagi.atlassian.net/browse/USH-4069
[Formplayer PR](https://github.com/dimagi/formplayer/pull/1549)

The Formplayer PR adds the module name to the request body. This HQ PR uses the module name for metric tagging. It also needs to remove the tag specific parameters before the search configuration/filters are extracted from the request dictionary.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
[Detailed Tagging](https://www.commcarehq.org/hq/flags/edit/detailed_tagging/)
New feature flag `HIGH_COUNT_DETAILED_TAGGING`

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
locally tested and tested on staging.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Tests exist that parameters are correctly extracted from request dictionary. I added to the test to ensure the keys specific to tagging are removed.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
no QA

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->
Related Formplayer PR must be reverted first. The Formplayer PR adds specific tag keys to the request and this HQ PR removes that tag specific key from the request_dict in `extract_search_request_config`. This will not be handled correctly if only the HQ PR is reverted.

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

Jtang-1 committed:
feat: collects metric tag from request body and adds to metric

Jtang-1 committed:
feat: removes tag parameters from request_dict used for getting case search results

Jtang-1 committed:
add tests and handle removing tag to lower level function

Jtang-1 committed:
feat: gates certain tags behind feature flag to control costs

snopoke commented on corehq/util/metrics/const.py:
```diff
@@ -15,3 +15,9 @@
+MODULE_NAME_TAG = "module_name"
+
+# These tags will only be added for domains with FF DETAILED_TAGGING enabled
+GATED_DETAILED_TAGS = (MODULE_NAME_TAG)
```
convert string to list (I assume this was meant to be a tuple but was missing a comma)
```suggestion
GATED_DETAILED_TAGS = [MODULE_NAME_TAG]
```

Jtang-1 replied:
yes it was, thanks. Ended up making it a set [1fd9571](https://github.com/dimagi/commcare-hq/pull/34226/commits/1fd9571e4777d2e43734248043e55e29f4768088)

snopoke reviewed (commented):
This is going to balloon the metric count in DD.

snopoke commented:
> @snopoke do you recommend creating another FF (separate from https://www.commcarehq.org/hq/flags/edit/detailed_tagging/) for this even more detailed tagging?

That could be a good idea otherwise this will apply to the 82 domains in that flag.

Jtang-1 commented on corehq/util/metrics/__init__.py:
```diff
@@ -326,6 +326,17 @@ def limit_domains(domain_name):
 
+def limit_tags(tags: Dict[str, str], domain: str):
+    from corehq import toggles
+    if toggles.DETAILED_TAGGING.enabled(domain):
```
For `limit_domains`, are you thinking adding to the allowed domains or restricting to only domains with the FF. I'm guessing the latter? Do you if there was a reason for the current limited domains?

Side note: Simon suggested creating a FF separate from `DETAILED_TAGGING` for the module name tagging.

```
if not settings.IS_SAAS_ENVIRONMENT:
    return domain_name
if domain_name and domain_name in _domains_to_tag():
    return domain_name
if toggles.DETAILED_TAGGING.enabled(domain_name):
    return domain_name
return '__other__'
```

or 
```
if toggles.DETAILED_TAGGING.enabled(domain_name):
    return domain_name
return '__other__'
    ```

esoergel replied:
> restricting to only domains with the FF

^ this.  The existing implementation is just something I did arbitrarily - I wanted USH domains tagged and didn't see a good way to do it, and so tagging enterprise domains seemed reasonable (as those generally get the most attention from the tech team).  However I now think that's overbroad.

This seems fine:
```python
if toggles.DETAILED_TAGGING.enabled(domain_name):
    return domain_name
return '__other__'
```

Jtang-1 replied:
[8cbe81d](https://github.com/dimagi/commcare-hq/pull/34226/commits/8cbe81d45eb5c9c4dc88f6b90b953650bc9042bf)

Jtang-1 committed:
creates new feature flag as to not balloon metric counts

Existing FF already contains many domains that do not need the module_name
tagging.

Jtang-1 committed:
bug: should be an iterable

Jtang-1 committed:
style: following existing standard for non-filter keys

esoergel commented on corehq/util/metrics/const.py:
```diff
@@ -15,3 +15,9 @@
+# Datadog tags
+MODULE_NAME_TAG = "module_name"
+
+# These tags will only be added for domains with FF DETAILED_TAGGING enabled
```
```suggestion
# These tags will only be added for domains with FF HIGH_COUNT_DETAILED_TAGGING enabled
```

esoergel commented on corehq/util/metrics/__init__.py:
```diff
@@ -326,6 +326,17 @@ def limit_domains(domain_name):
+    if toggles.HIGH_COUNT_DETAILED_TAGGING.enabled(domain):
+        return tags
+
+    for key in list(tags.keys()):
+        if key in GATED_DETAILED_TAGS:
+            del tags[key]
+    return tags
```
Nit: I'd prefer returning a new dict versus modifying the existing one, just to avoid mutation, which can be tricky to reason about.

```suggestion
    return {k: v for k, v in tags.items() if k not in GATED_DETAILED_TAGS}
```

esoergel reviewed (approved):
couple nitpicky comments, but nothing else blocking

Jtang-1 committed:
explicitly tag only domains that have FF enabled
The previous implementation was arbitrary, the underlying goal was to tag USH domains.

Jtang-1 committed:
create new dict to avoid mutation

Jtang-1 committed:
correct comment

Jtang-1 committed:
Merge branch 'master' into jt/module-tag-app-aware-search

esoergel commented on corehq/util/metrics/__init__.py:
```diff
@@ -319,9 +320,7 @@ def limit_domains(domain_name):
     """
-    if not settings.IS_SAAS_ENVIRONMENT:
-        return domain_name
-    if domain_name and domain_name in _domains_to_tag():
```
this is the only usage of `_domains_to_tag`, so that can be removed now too (non-blocking)

Jtang-1 replied:
[c88f5fc](https://github.com/dimagi/commcare-hq/pull/34226/commits/c88f5fccc9c07a58f5ea60d8ace89c147784517b)

Jtang-1 committed:
deletes unused function

Jtang-1 committed:
bug: error importing toggles in init file

Jtang-1 committed:
style: rename constant to better support future additional tags
